zebra-network/src/peer/handshake.rs

```diff
@@ -464,19 +464,39 @@ where
                                 let (tx, rx) = oneshot::channel();
                                 let request = Request::Ping(Nonce::default());
                                 tracing::trace!(?request, "queueing heartbeat request");
-                                if server_tx
-                                    .send(ClientRequest {
-                                        request,
-                                        tx: tx.into(),
-                                        span: tracing::Span::current(),
-                                    })
-                                    .await
-                                    .is_err()
-                                {
-                                    tracing::trace!(
-                                        "error sending heartbeat request, shutting down"
-                                    );
-                                    return;
+                                match server_tx.try_send(ClientRequest {
+                                    request,
+                                    tx: tx.into(),
+                                    span: tracing::Span::current(),
+                                }) {
+                                    Ok(()) => {
+                                        match server_tx.flush().await {
+                                            Ok(()) => {}
+                                            Err(e) => {
+                                                // TODO: we can't get the client request for this failure,
+                                                // so we can't ensure the invariant holds
+                                                panic!("flushing client request failed: {:?}", e);
+                                            }
+                                        }
+                                    }
+                                    Err(e) => {
+                                        tracing::trace!(
+                                            ?e,
+                                            "error sending heartbeat request, shutting down"
+                                        );
+                                        if e.is_disconnected() {
+                                            let ClientRequest { tx, .. } = e.into_inner();
+                                            let _ =
+                                                tx.send(Err(PeerError::ConnectionClosed.into()));
+                                        } else if e.is_full() {
+                                            let ClientRequest { tx, .. } = e.into_inner();
+                                            let _ = tx.send(Err(PeerError::Overloaded.into()));
```

Comment:
We close the connection if the queue is full, rather than waiting for it to empty.

Here are our options for fixing this issue:
* can we await the queue being empty?
* can we just ignore this heartbeat, and pick it up on the next request?
  * it might be better to close the connection, to avoid hangs   
* is it ok to just close the connection if the queue is full?

Reply:
Thinking about this some more…

The old behaviour was buggy, because it could hang forever waiting for the queue to be ready.
The new behaviour is better, because it doesn't hang, but it also churns connections more than needed.
Ideally, we should wait for the queue to be ready with a timeout, and close the connection if it times out.

Reply:
Since the latest code in this PR works, I'd be happy to make the timeout change in a follow-up PR. I don't think it's a particularly high priority, unless it really affects performance.

Reply:
See #1551